### ISSUE TITLE
Gating upstream `tree-diff` behind an opt-in flag

### ISSUE THREAD

**Issue by jasagredo** [stevana/quickcheck-state-machine#30]
It turns out that a lot of code (at least for us in IOG) depends on the hackage `tree-diff`, and we have lots of instances for it. I see there is an issue with the licenses, perhaps because you may want QSM to be usable on closed-source environments?

However how would you feel about gating the dependency behind a flag, such that users of QSM could decide whether they want to use the upstream `tree-diff` (and be GPL compliant) or the in-tree one? It would basically consist on:

```cabal
flag vendored-tree-diff
  default: True
  manual: True

library 
  ...
  if flag(vendored-tree-diff)
  then
    exposed-modules: 
      Test.StateMachine.TreeDiff
      Test.StateMachine.TreeDiff.Class
      Test.StateMachine.TreeDiff.Expr
      Test.StateMachine.TreeDiff.List
      Test.StateMachine.TreeDiff.Pretty
      Test.StateMachine.TreeDiff.Tree
    build-depends:
      ... -- dependencies of tree-diff
  else
    build-depends:
      tree-diff
    mixins:
      tree-diff (Data.TreeDiff.Class as Test.StateMachine.TreeDiff, ...)
```

This way QSM would remain BSD-3 but users who want to keep up with the upstream tree-diff can still opt-in for it. It kind of emulates the way it works in GHC with gmp and bignum.

I have not tried the above, but I suspect there would be a way to make it work.

@stevana (pinging because it is a question directly for you more than an issue to solve eventually)

**Comment by stevana** [comment]
I don't understand the relationship between licenses either.

I've seen that for example Hakyll is BSD licensed, dispite depending on pandoc (which is GPL), there's an issue in their bug tracker where some people argue this isn't ok, while others say it's fine (but the resulting binary will be GPL).

I think the best thing would be to if somebody reimplemented tree-diff without looking at the source code and license it as BSD (again).

**Comment by jasagredo** [comment]
This was solved by #31
